src/witan/plot/series.clj

```diff
@@ -0,0 +1,193 @@
+
+  The rf oprates over a series of:
+  {:series-key foo :x-val calendar-year} {:actual population}
+  {:series-key foo :x-val calendar-year} {:low-95 :high-95 :iqr3 :iqr1 :median}
+
+  and produces solid lines for actual and dashed lines and ribbon
+  plots of 50 and 25 alpha for median/iqr/95"
+  [{::keys [series-specs domain-map]}]
+  (fn
+    ([]
+     {::series-specs series-specs
+      ::domain-map domain-map
+      ::series (sorted-map)
+      ::legend-spec (sorted-set)})
+    ([{::keys [series legend-spec]}]
+     {::legend-spec (into []
+                          (map (fn [series-key]
+                                 (let [spec (series-specs series-key)]
+                                   (try
+                                     [:shape (:label spec series-key)
+                                      {:color (:color spec)
+                                       :shape (:legend-shape spec)
+                                       :size 15
+                                       :stroke {:size 4.0}}]
+                                     (catch Exception e
+                                       (throw (ex-info "Couldn't create legend"
+                                                       {:spec spec :series-key series-key}
+                                                       e)))))
+                                 ))
+                          legend-spec)
+      ::series (into []
+                     (mapcat (fn [s] (collected-data->series s series-specs domain-map)))
+                     series)})
+    ([acc [{::keys [series-key x-key]} v]]
+     (-> acc
+         (update-in [::series series-key] collect-data x-key v)
+         (update ::legend-spec (fnil conj []) series-key)))))
+
+
+(defn line-ci-series-missing-is-0 [{::keys [data series-key x-key actual-key series-specs domain-map]}]
+  (transduce
+   (plot-friendly-kv-xf series-key x-key actual-key
+                        ;; :academic-year :calendar-year :population
+                        )
+   (line-ci-series-missing-is-0-rf-f
+    {::series-specs series-specs ;; ay-cs
+     ::domain-map domain-map #_(into (sorted-map)
+                                     (sequence
+                                      (map (fn [y z] [y z]))
+                                      (range 2015 2021)
+                                      (repeat 0)))})
+   data))
+
+(comment
+
+  (def medians-example
+    {:low-95pc-bound  {2020 1
+                       2021 2
+                       2022 3}
+     :q1              {2020 3
+                       2021 3
+                       2022 3}
+     :median          {2020 5
+                       2021 5
+                       2022 5}
+     :q3              {2020 7
+                       2021 7
+                       2022 7}
+     :high-95pc-bound {2020 9
+                       2021 9
+                       2022 9}})
+
+  (def actuals-example
+    {:actual {2020 5
+              2021 5
+              2022 5}})
+  )
```

Comment:
Does this need to stay?

Reply:
yeah, as those are the target data structures and I always end up trying to figure them out again. I need to turn them into tests and clojure.spec, but I'm not sure I'm happy enough with them yet.